malwarefront/src/components/RelationsPlot.js

```diff
@@ -1,233 +1,187 @@
-        let type = nodeInfo.type;
-
-        let typeMapping = {
-            file: "sample",
-            config: "config",
-            static_config: "config",
-            text_blob: "blob"
+    const addNodes = (prevNodesState, newNode, newEdge = false) => {
+        const getNode = (hash) => {
+            return prevNodesState.nodes.find((n) => n.id === hash);
         }
 
-        let styleMapping = {
-            sample: "bg-danger",
-            config: "bg-success",
-            blob: "bg-info"
-
+        const getEdge = (edge) => {
+            return prevNodesState.edges.find((e) => e.parent === edge.parent && e.child === edge.child)
         }
 
-        type = typeMapping[type];
-        let d = new Date(date)
-        let html = "";
-
-        if (!expanded) {
-            html = `<div><div class='mainNode'> 
-                        <div class="card" style="width:13rem; cursor: pointer;">
-                            <div class="card-header node-header-active ${styleMapping[type]}" style="padding-top: 11px; padding-bottom: 11px;">
-                             ${capitalize(type)} <span class="date">${d.toLocaleDateString()}</span></div>
-                              <div class="card-body">
-                                  <p class="card-text">
-                                      <small class="text-muted">
-                                            <a href="/${type}/${hash}">
-                                                ${hash.substr(0,16)}
-                                            </a>
-                                      </small>
-                                  </p>
-                              </div>
-                         </div>
-                    <div class="card-footer bg-transparent tags" style="max-width: 13rem; white-space:pre-wrap">`;
+        let newNodesState = {}
+        let existingNode = getNode(newNode.id)
+        if(!existingNode)
+        {
+            // Append new node
+            if (!newEdge)
+                newNode = convertObject(newNode, true)
+            else
+                newNode = convertObject(newNode, false)
+            newNodesState.nodes = [...prevNodesState.nodes, newNode]
         } else {
-            html = `<div><div class='mainNode'> 
-                            <div class="card" style="width:13rem;">
-                             <div class="card-header node-header-expanded ${styleMapping[type]}">${capitalize(type)} <span class="date">${d.toLocaleDateString()}</span></div>
-                              <div class="card-body">
-                                  <a href="/${type}/${hash}">${hash.substr(0,16)}</a><br></div>
-                              </div>
-                         </div>
-                    <div class="card-footer bg-transparent tags" style="max-width: 13rem; white-space:pre-wrap">`;
+            // Update node in the same place
+            if ((!existingNode.expanded && !newEdge) || existingNode.expanded)
```

Comment:
I think it could be simplified a bit e.g.:
```javascript
if(existingNode.expanded) {
    // If node is expanded: it never would be updated
    newNodesState.nodes = [...prevNodesState.nodes]
} else {
    // Node update assuming that !existingNode.expanded
    ...
}
```